File: OOP3Behav4.cpp
#include "pt4.h"
using namespace std;

class Validator
{
public:
    virtual string Validate(string s);
};

string Validator::Validate(string s)
{
    return "";
}

// Implement the EmptyValidator, NumberValidator
//   and RangeValidator descendant classes
class EmptyValidator : public Validator
{
public:
    string Validate(string s) override
    {
        if (s.empty())
            return "!Empty text";
        else
            return "";
    }
};

class NumberValidator : public Validator
{
public:
    string Validate(string s) override;
};

string NumberValidator::Validate(string s)
{
    if (s.empty())
        return string("!'") + s + "': not a number";
    for (int i = 0; i < s.length(); i++)
        if (!isdigit(s[i]) && (i || s[i] != '-'))
            return string("!'") + s + "': not a number";
    return "";
}

class RangeValidator : public Validator
{
public:
    RangeValidator(int a, int b)
    {
        minInt = min(a, b);
        maxInt = max(a, b);
    }
    string Validate(string s) override;

private:
    int minInt, maxInt;
};

string RangeValidator::Validate(string s)
{
    if (s.empty())
        return string("!'") + s + "': not in range " + to_string(minInt) + ".." + to_string(maxInt);
    for (int i = 0; i < s.length(); i++)
        if (!isdigit(s[i]) && (i || s[i] != '-'))
            return string("!'") + s + "': not in range " + to_string(minInt) + ".." + to_string(maxInt);
    int n = stoi(s);
    if (n < minInt || n > maxInt)
        return string("!'") + s + "': not in range " + to_string(minInt) + ".." + to_string(maxInt);
    return "";
}

// Implement the TextBox and TextForm classes
class TextBox
{
public:
    TextBox() : text("") { v = new Validator(); }
    ~TextBox() { delete v; }
    void setText(string text) { this->text = text; }
    void setValidator(Validator *v)
    {
        delete this->v;
        this->v = v;
    }
    string Validate() { return v->Validate(text); }
    string Validate() const { return v->Validate(text); }

private:
    string text;
    Validator *v;
};

class TextForm
{
public:
    TextForm(int n) { tb.resize(n); }
    void SetText(int ind, string text) { tb[ind].setText(text); }
    void SetValidator(int ind, Validator *v) { tb[ind].setValidator(v); }
    string Validate()
    {
        string s;
        for (const auto &t : tb)
            s += t.Validate();
        return s;
    }

private:
    vector<TextBox> tb;
};
void Solve()
{
    Task("OOP3Behav4");
    int n, a, b, k;
    pt >> n >> a >> b >> k;
    vector<pair<int, char>> v;
    for (int i = 0; i < k; i++)
    {
        int in;
        char c;
        pt >> in >> c;
        v.push_back({in, c});
    }
    for (int i = 0; i < 5; i++)
    {
        TextForm t(n);
        string s;
        for (int i = 0; i < n; i++)
        {
            pt >> s;
            t.SetText(i, s);
        }
        for (auto &p : v)
        {
            int in = p.first;
            char c = p.second;
            if (c == 'E')
                t.SetValidator(in, new EmptyValidator());
            else if (c == 'N')
                t.SetValidator(in, new NumberValidator());
            else
                t.SetValidator(in, new RangeValidator(a, b));
        }
        pt << t.Validate();
    }

}

File: OOP3Behav4.cs
// File: "OOP3Behav4"
using PT4;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PT4Tasks
{
    public class MyTask : PT
    {
        public class Validator
        {
            public virtual string Validate(string s)
            {
                return "";
            }
        }

        // Implement the EmptyValidator, NumberValidator
        //   and RangeValidator descendant classes
        public class EmptyValidator : Validator
        {
            public override string Validate(string s)
            {
                if (s.Length == 0)
                {
                    return "!Empty text";
                }
                return "";
            }
        }
        public class NumberValidator : Validator
        {

            public override string Validate(string s)
            {
                int num;
                bool isNum = int.TryParse(s, out num);
                if (isNum)
                {
                    return "";
                }
                return "!'" + s + "': not a number";
            }
        }
        public class RangeValidator : Validator
        {
            int min, max;
            public RangeValidator(int a, int b)
            {
                if (a < b)
                {
                    min = a; max = b;
                }
                else
                {
                    min = b; max = a;
                }
            }
            public override string Validate(string s)
            {
                int num;
                bool isNum = int.TryParse(s, out num);
                if (isNum && num >= min && num <= max)
                {
                    return "";
                }
                return "!'" + s + "': not in range " + min + ".." + max;
            }
        }
        // Implement the TextBox and TextForm classes

        public class TextBox
        {
            string text;
            Validator v;
            public TextBox()
            {
                v = new Validator();
                text = "";
            }
            public void SetText(string text)
            {
                this.text = text;
            }
            public void SetValidator(Validator v)
            {
                this.v = v;
            }
            public string Validate()
            {
                return v.Validate(text);
            }
        }

        public class TextForm
        {
            TextBox[] tb;

            public TextForm(int N)
            {
                tb = new TextBox[N];
                for (int i = 0; i < N; i++)
                    tb[i] = new TextBox();
            }
            public void SetText(int ind, string text)
            {
                tb[ind].SetText(text);
            }
            public void SetValidator(int ind, Validator v)
            {
                tb[ind].SetValidator(v);
            }
            public string Validate()
            {
                string str = "";
                for (int i = 0; i < tb.Length; i++)
                    str += tb[i].Validate();
                return str;
            }
        }

        public static void Solve()
        {
            Task("OOP3Behav4");
            int n = GetInt();
            int a = GetInt();
            int b = GetInt();

            int k = GetInt();
            int[] arri = new int[k];
            char[] c = new char[k];

            for (int i = 0; i < k; i++)
            {
                arri[i] = GetInt();
                c[i] = GetChar();
            }

            string[] s1 = new string[n];
            for (int i = 0; i < n; i++)
            {
                s1[i] = GetString();
            }

            string[] s2 = new string[n];
            for (int i = 0; i < n; i++)
            {
                s2[i] = GetString();
            }

            string[] s3 = new string[n];
            for (int i = 0; i < n; i++)
            {
                s3[i] = GetString();
            }

            string[] s4 = new string[n];
            for (int i = 0; i < n; i++)
            {
                s4[i] = GetString();
            }

            string[] s5 = new string[n];
            for (int i = 0; i < n; i++)
            {
                s5[i] = GetString();
            }

            TextForm tf = new TextForm(n);

            for (int i = 0; i < k; i++)
            {
                if (c[i] == 'E')
                    tf.SetValidator(arri[i], new EmptyValidator());
                else if (c[i] == 'R')
                    tf.SetValidator(arri[i], new RangeValidator(a, b));
                else
                    tf.SetValidator(arri[i], new NumberValidator());
            }

            for (int i = 0; i < n; i++)
            {
                tf.SetText(i, s1[i]);
            }
            Put(tf.Validate());
            for (int i = 0; i < n; i++)
            {
                tf.SetText(i, s2[i]);
            }
            Put(tf.Validate());
            for (int i = 0; i < n; i++)
            {
                tf.SetText(i, s3[i]);
            }
            Put(tf.Validate());
            for (int i = 0; i < n; i++)
            {
                tf.SetText(i, s4[i]);
            }
            Put(tf.Validate());
            for (int i = 0; i < n; i++)
            {
                tf.SetText(i, s5[i]);
            }
            Put(tf.Validate());
        }
    }
}
